### PR TITLE
perf: Evaluate zerofier efficiently

### PR DESCRIPTION
**Pull request by pablodeymo**

### CODE REVIEW

File: src/air/constraints/evaluator.rs [hunk 1]
```diff
@@ -153,16 +153,20 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
...
 
-        let mut zerofier_evaluations = evaluate_polynomial_on_lde_domain(
-            &x_n_1,
-            domain.blowup_factor,
-            domain.interpolation_domain_size,
-            &domain.coset_offset,
+        let offset = FieldElement::<F>::from(self.air.context().options.coset_offset);
+        let offset_pow = offset.pow(trace_length);
+        let one = FieldElement::<F>::one();
+        let mut zerofier_evaluations = get_powers_of_primitive_root_coset(
```
Comment (Oppen) [inline src/air/constraints/evaluator.rs:161]: For a different PR, but I'd suggest `get_powers_of_primitive_root_coset` to return the iterator. Or, at least, to implement an `iter_powers_of_primitive_root_coset` and implement the former in terms of that and a `collect`. This avoids creating a `Vec<_>` that we only iterate once.

Comment (diegokingston) [reply src/air/constraints/evaluator.rs:161]: We can do that in another PR

File: src/air/constraints/evaluator.rs [hunk 2]
```diff
@@ -153,16 +153,20 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
             })
             .collect();
 
-        let x_n = Polynomial::new_monomial(FieldElement::<F>::one(), trace_length);
-        let x_n_1 = x_n - FieldElement::<F>::one();
+        let blowup_factor_log2 = u64::from(blowup_factor.trailing_zeros());
```
[the remark this answers is not in the document]
Comment (diegokingston) [reply src/air/constraints/evaluator.rs:156]: We can change the name to blowup_factor_order or something like that

File: src/air/constraints/evaluator.rs [hunk 3]
```diff
@@ -171,6 +175,7 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
                 .map(|row| {
                     zerofier_evaluations
                         .iter()
+                        .cycle()
```
Comment (Oppen) [inline src/air/constraints/evaluator.rs:178]: I don't understand how this `cycle` works here or what works differently now. Could you explain a bit further?

File: src/air/constraints/evaluator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pablodeymo) [reply src/air/constraints/evaluator.rs]: I've just changed the variable name!

**Review by Oppen — APPROVED** [review]

**Review by entropidelic — APPROVED** [review]

**Review by MauroToscano — CHANGES_REQUESTED** [review]
We need to measure memory improvements with dhat, since there are no speedups

**Pull request merged**